### DISCUSSION TITLE
Feature proposal: ability to declare all dependencies in Gemfile, not in gemspec

### DISCUSSION THREAD

**Discussion by jeromedalbert** [ruby/rubygems#7869] [Ideas]
In a non-gem Rails app, all the dependencies are declared in the Gemfile. Nice and tidy.
But in a gem, dependencies nowadays are often declared in 2 different files at the same time: `Gemfile` for dev dependencies, and `my_gem.gemspec` for runtime dependencies.

Wouldn't it be nice if all of a gem's dependencies could be declared in the Gemfile? Or is this a crazy idea?

I don't often create new gems but when I do, I wish I could declare all of my dependencies in the Gemfile. Gemfiles are nice. I like them. Declaring dependencies in them feels less clunky than in a gemspec.

Gemfiles can call a `gemspec` method to add dependencies from the gemspec, among other things.
Conversely, why can't gemspecs call a `gemfile` (or similar) method to add dependencies from the Gemfile?

It could look something like this:

```ruby
# my_gem.gemspec

Gem::Specification.new do |spec|
  # ...

  spec.add_gemfile_dependencies
end
```

or maybe just

```ruby
Gem::Specification.new do |spec|
  # ...

  spec.gemfile
end
```

or if you want to include specific Bundler groups:

```ruby
Gem::Specification.new do |spec|
  # ...

  spec.add_gemfile_dependencies(:default, :release) # or whatever groups you want
end
```

This would be optional behavior and wouldn't break anything. Here is a hacky but working proof of concept: [gist](https://gist.github.com/jeromedalbert/ebc5af3b0b5a28d186e0629f62cb825d).

Would anyone find this useful, or is it just me?

**Comment by indirect** [comment]
The main reason is historical! For a long time, Bundler was not able to make changes to RubyGems itself, and had to be an additive layer on top. In that situation, gem dependencies had to live exclusively in the gemspec where RubyGems could find them.

Today, it seems like this would be pretty doable! I'd be open to a PR adding this functionality.

**Comment by simi** [comment]
I don't like this idea spreading runtime dependencies across multiple files. You should do the opposite way, specify runtime deps in gemspec and use `gemspec` in `Gemfile`. IMHO we should finally write guide how to compose your library dependencies in modern way and ensure everything provided by RubyGems/Bundler follows this like `bundle gem` and RubyGems maintained gems.

**Reply by jeromedalbert** [reply]
So you're saying the current way `bundle gem` generates gems (with dev dependencies in Gemfile and runtime dependencies in gemspec) is not best? You would want everything in the `gemspec`, which would contain both `add_development_dependency` and `add_dependency`?

**Reply by deivid-rodriguez** [reply]
@simi My understanding is that what you are suggesting is what people do nowadays, but it's true that it seems more convenient to have all the dependencies of your gem specified in a single place?

I've historically recommended development dependencies to be specified in the Gemfile, because of it's flexibility. For example, you can use `git` sources or other Bundler niceties that are not available through `add_developement_dependency`. But it's also true that if you specify development dependencies in the Gemfile only, then they are not visible to RubyGems.org, so they don't show up as "Development Dependencies" in the gem page (like in https://rubygems.org/gems/compact_index, for example), which is neat.

Maybe an idea like @jeromedalbert's one can bring the best of both worlds and also allows you to always keep your dependencies in a single place, the `Gemfile`.
